components/camel-undertow/src/main/java/org/apache/camel/component/undertow/handlers/CamelMethodHandler.java

```diff
@@ -16,76 +16,127 @@
+            } else {
+                exchange.setStatusCode(StatusCodes.METHOD_NOT_ALLOWED);
+                exchange.getResponseHeaders().put(ExchangeHeaders.CONTENT_TYPE, MimeMappings.DEFAULT_MIME_MAPPINGS.get("txt"));
+                exchange.getResponseHeaders().put(ExchangeHeaders.CONTENT_LENGTH, 0);
+                exchange.endExchange();
+            }
         }
     }
 
-    public synchronized void add(String[] methods, HttpHandler handler) {
-        Map<String, HttpHandler> adding = new HashMap<String, HttpHandler>();
-        for (String method : methods) {
-            adding.put(method, handler);
+    public HttpHandler add(String methods, HttpHandler handler) {
+        HttpHandler result = null;
+        synchronized (methodMap) { // we lock on methodMap to get a reliable sum of refCounts in remove(String)
+            for (String method : splitMethods(methods)) {
+                MethodEntry en = methodMap.computeIfAbsent(method, m -> new MethodEntry());
+                result = en.addRef(handler, method);
+            }
         }
-        methodMap.putAll(adding);
         handlerString = null;
+        return result;
     }
 
-    public synchronized void remove(String[] methods) {
-        for (String method : methods) {
-            methodMap.remove(method);
+
+    public boolean remove(String methods) {
+        boolean result;
+        synchronized (methodMap) { // we lock on methodMap to get a reliable sum of refCounts
+            for (String method : splitMethods(methods)) {
+                final MethodEntry en = methodMap.get(method);
+                if (en != null) {
+                    en.removeRef();
+                }
+            }
+            result = methodMap.values().stream().mapToInt(en -> en.refCount).sum() == 0;
         }
         handlerString = null;
+        return result;
     }
 
-    public synchronized void addDefault(HttpHandler handler) {
-        if (defaultHandler != null) {
-            throw new IllegalArgumentException(String.format(
-                "Duplicate default handler: '%s', '%s'", defaultHandler, handler));
+    public String toString() {
+        if (handlerString == null) {
+            handlerString = "CamelMethodHandler[" + methodMap + "]";
         }
-        defaultHandler = handler;
-        handlerString = null;
+        return handlerString;
     }
 
-    public synchronized void removeDefault() {
-        defaultHandler = null;
-        handlerString = null;
+    private String[] splitMethods(String methods) {
+        String[] result = methods != null ? methods.split(",") : DEFAULT_METHODS;
+        return result.length == 0 ? DEFAULT_METHODS : result;
     }
 
-    public boolean isEmpty() {
-        return defaultHandler == null && methodMap.isEmpty();
-    }
+    static class MethodEntry {
+        /**
+         * The number of references pointing to {@link #handler}
+         */
+        private int refCount;
```

Comment:
Add a bit of empty lines so this code is formatted nice between fields, constructor, and methods

Reply:
Fixed in a85b3e3